GVFS/GVFS.Common/Tracing/JsonTracer.cs

```diff
@@ -339,40 +340,39 @@ private void WriteEvent(string eventName, EventLevel level, Keywords keywords, E
 
         private void MarkAndLogListenerRecovery(EventListener recoveredListener)
         {
+            // Check ContainsKey first (rather than always calling TryRemove) because ContainsKey
+            // is lock-free and recoveredListener should rarely be in failedListeners
             if (!this.failedListeners.ContainsKey(recoveredListener))

```

Comment:
@wilbaker and I talked offline and it's fine if the ContainsKey check returns true when it's turned to false (no-lock).  The TryRemove would then catch it.

What about the case when Contains key is false but becomes true.  Is that a case that could happen and if so would it be handled?

Reply:
If `ContainsKey` returns `false` but another thread *just* added `recoveredListener` it means the listener is failing and succeeding on two different threads at the same time.

That means the thread that just added the listener to `failedListeners` will log the failure and this thread (where `ContainsKey` returned false) will exit early and not return anything.

Then the next time `MarkAndLogListenerRecovery` is called `ContainsKey` will return true and recovery will be logged.

Note that this race could happen even if we removed the call to `ContainsKey` and relied on `TryRemove`.

Thread 1: Enters `MarkAndLogListenerFailure`
Thread 2: Enters `MarkAndLogListenerRecovery`
Thread 2: Calls `TryRemove` and it returns false (just as `ContainsKey` would)
Thread 1: Calls `TryAdd` and it returns true
Thread 1: Logs listener error
Later: `MarkAndLogListenerRecovery` is called and recovery is logged

The only change introduced by `ContainsKey` is that if it crosses with "Thread 1: Calls `TryAdd`" `ContainsKey` could return either `true` or `false`.

In general if we're getting failures and successes at the same time on two different threads, there's not a guarantee on how the failure/recovery messages will be logged.